// react-native-auth0/lib/commonjs/auth/authError.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.handleInvalidToken = exports.default = void 0;
var _baseError = _interopRequireDefault(require("../utils/baseError"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
class AuthError extends _baseError.default {
  constructor(response) {
    const {
      status,
      json,
      text
    } = response;
    const {
      error,
      error_description: description
    } = json ?? {
      error: undefined,
      error_description: undefined
    };
    super(error || 'a0.response.invalid', description || text || handleInvalidToken(response) || 'unknown error');
    this.json = json;
    this.status = status;
  }
}
exports.default = AuthError;
const handleInvalidToken = response => {
  var _response$headers;
  return response !== null && response !== void 0 && (_response$headers = response.headers) !== null && _response$headers !== void 0 && (_response$headers = _response$headers.get('www-authenticate')) !== null && _response$headers !== void 0 && _response$headers.match(/error="invalid_token"/g) ? 'invalid_token' : null;
};
exports.handleInvalidToken = handleInvalidToken;
